File: chrome/android/modules/stack_unwinder/internal/java/src/org/chromium/chrome/modules/stack_unwinder/StackUnwinderModuleContentsImpl.java
// Copyright 2020 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.modules.stack_unwinder;

import org.jni_zero.NativeMethods;

import org.chromium.build.annotations.NullMarked;

/**
 * Provides access to the stack unwinder native code functions within the dynamic feature module.
 */
@NullMarked
public class StackUnwinderModuleContentsImpl implements StackUnwinderModuleContents {

    @Override
    public long getDoNothingFunction() {
        return StackUnwinderModuleContentsImplJni.get().getDoNothingFunction();
    }

    @NativeMethods("stack_unwinder")
    interface Natives {
        long getDoNothingFunction();
    }
}

File: chrome/android/modules/stack_unwinder/internal/stack_unwinder_module_contents_impl.cc
// Copyright 2020 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <memory>
#include <type_traits>
#include <utility>

#include "chrome/android/features/stack_unwinder/public/function_types.h"

// Must come after all headers that specialize FromJniType() / ToJniType().
#include "chrome/android/modules/stack_unwinder/internal/jni_headers/StackUnwinderModuleContentsImpl_jni.h"

void DoNothing() {
  return;
}

static_assert(std::is_same<stack_unwinder::DoNothingFunction,
                           decltype(&DoNothing)>::value,
              "DoNothingFunction typedef must match the declared function "
              "type");

static jlong JNI_StackUnwinderModuleContentsImpl_GetDoNothingFunction(
    JNIEnv* env) {
  return reinterpret_cast<jlong>(&DoNothing);
}

DEFINE_JNI(StackUnwinderModuleContentsImpl)
